File: Hash/169-majorityElement/169.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
private:
    unordered_map<int, int> num_count;
public:
    int majorityElement(vector<int>& nums) {
        int limit = nums.size() / 2;
        for (int &num : nums) {
            if (num_count.find(num) != num_count.end()) {
                ++num_count[num];
            } else {
                num_count[num] = 1;
            }
            if (num_count[num] > limit) {
                return num;
            }
        }
        return 0;
    }
};

File: Hash/169-majorityElement/go/169.go
package main

import "math/rand"

func majorityElement(nums []int) int {
	numCount := make(map[int]int)
	var limit int = len(nums) / 2
	for _, num := range nums {
		numCount[num]++
		if numCount[num] > limit {
			return num
		}
	}
	return 0
}

// 随机化（神奇的方法）
func majorityElement_2(nums []int) int {
	limit := len(nums) / 2
	for {
		randID := rand.Intn(len(nums))
		candidate := nums[randID]
		count := 0
		for _, num := range(nums) {
			if num == candidate {
				count++
			}
		}
		if count > limit {
			return candidate
		}
	}
}